app/views/contexts/section_summary_context.py

```diff
@@ -113,18 +113,42 @@ def _title_for_location(self, location):
         )
         return title
 
-    def _custom_summary_elements(self, section_summary, current_location, section):
+    def _custom_summary_elements(
+        self, section_summary, current_location, section, routing_path
+    ):
         for summary_element in section_summary:
             if summary_element["type"] == "List":
                 yield self._list_summary_element(
-                    summary_element, current_location, section
+                    summary_element, current_location, section, routing_path
                 )
 
-    def _list_summary_element(self, summary, current_location, section) -> Mapping:
+    def _list_summary_element(
+        self, summary, current_location, section, routing_path
+    ) -> Mapping:
```

Comment:
You can remove the list comprehension and I think we should make `ListContext` a property. The add_link logic could stay as is if you fetch the first block at the end.

I think we need to split this method up as it is doing a lot of things. And I think we can do that by generating the routing path on `__init__`. At the moment, it is passed to every method, I don't see why we can generate the routing path on the initialisation of the object, then you don't have to pass the routing path around to all the different methods.


```py
    def _list_summary_element(
        self, summary, current_location, section, routing_path
    ) -> Mapping:
        list_collector_block = None
        edit_block_id, remove_block_id, primary_person_edit_block_id = None, None, None
        current_list = self._list_store[summary["for_list"]]

        list_collector_blocks = list(
            self._schema.get_list_collectors_for_list(
                section, for_list=summary["for_list"]
            )
        )

        list_collector_blocks_on_path = [
            list_collector_block
            for list_collector_block in list_collector_blocks
            if list_collector_block["id"] in routing_path.block_ids
        ]

        if list_collector_blocks_on_path:
            list_collector_block = list_collector_blocks_on_path[0]
            edit_block_id = list_collector_block["edit_block"]["id"]
            remove_block_id = list_collector_block["remove_block"]["id"]

        add_link = self._add_link(summary, section, list_collector_block)

        if len(current_list) == 1 and current_list.primary_person:
            primary_person_block = self._schema.get_list_collector_for_list(
                section, for_list=summary["for_list"], primary=True
            )

            primary_person_edit_block_id = primary_person_block["add_or_edit_block"][
                "id"
            ]
            edit_block_id = primary_person_block["add_or_edit_block"]["id"]

        rendered_summary = self._placeholder_renderer.render(
            summary, current_location.list_item_id
        )

        list_collector_block = list_collector_block or list_collector_blocks[0]

        list_context = ListContext(
            self._language,
            self._schema,
            self._answer_store,
            self._list_store,
            self._progress_store,
            self._metadata,
        )
        list_summary_context = list_context(
            list_collector_block["summary"],
            for_list=list_collector_block["for_list"],
            return_to="section-summary",
            edit_block_id=edit_block_id,
            remove_block_id=remove_block_id,
            primary_person_edit_block_id=primary_person_edit_block_id,
        )

        return {
            "title": rendered_summary["title"],
            "type": rendered_summary["type"],
            "add_link": add_link,
            "add_link_text": rendered_summary["add_link_text"],
            "empty_list_text": rendered_summary["empty_list_text"],
            "list_name": rendered_summary["for_list"],
            **list_summary_context,
        }
```